internal/app/machined/pkg/controllers/seccomp/seccomp.go

```diff
@@ -0,0 +1,169 @@
+
+// Inputs implements controller.StatsController interface.
+func (ctrl *ProfilesController) Inputs() []controller.Input {
+	return nil
+}
+
+// Outputs implements controller.StatsController interface.
+func (ctrl *ProfilesController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: seccomp.ProfileType,
+			Kind: controller.OutputExclusive,
+		},
+	}
+}
+
+// Run implements controller.StatsController interface.
+//
+//nolint:gocyclo,cyclop
+func (ctrl *ProfilesController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	// initially, wait for the machine-id to be generated and /var to be mounted
+	if err := r.UpdateInputs([]controller.Input{
+		{
+			Namespace: runtimeres.NamespaceName,
+			Type:      runtimeres.MountStatusType,
+			ID:        pointer.To(constants.EphemeralPartitionLabel),
+			Kind:      controller.InputWeak,
+		},
+	}); err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		_, err := r.Get(ctx, resource.NewMetadata(runtimeres.NamespaceName, runtimeres.MountStatusType, constants.EphemeralPartitionLabel, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				// in container mode EPHEMERAL is always mounted
+				if ctrl.V1Alpha1Mode != runtimetalos.ModeContainer {
+					// wait for the EPHEMERAL to be mounted
+					continue
+				}
+			} else {
+				return fmt.Errorf("error getting ephemeral mount status: %w", err)
+			}
+		}
+
+		break
+	}
+
+	// normal reconcile loop
+	if err := r.UpdateInputs([]controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
+			ID:        pointer.To(config.V1Alpha1ID),
+			Kind:      controller.InputWeak,
+		},
+	}); err != nil {
+		return err
+	}
+
+	r.QueueReconcile()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		cfg, err := r.Get(ctx, resource.NewMetadata(config.NamespaceName, config.MachineConfigType, config.V1Alpha1ID, resource.VersionUndefined))
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				continue
+			}
+
+			return fmt.Errorf("error getting config: %w", err)
+		}
+
+		cfgProvider := cfg.(*config.MachineConfig).Config()
+
+		touchedIDs := map[string]struct{}{}
+
+		for _, profile := range cfgProvider.Machine().SeccompProfiles() {
+			value, errJSON := json.Marshal(profile.Value())
+			if errJSON != nil {
+				return fmt.Errorf("error marshaling seccomp profile: %w", errJSON)
+			}
+
+			if err = r.Modify(ctx, seccomp.NewSeccompProfile(profile.Name()), func(res resource.Resource) error {
+				*res.(*seccomp.Profile).TypedSpec() = seccomp.ProfileSpec{
+					Name:  profile.Name(),
+					Value: profile.Value(),
+				}
+
+				return nil
+			}); err != nil {
+				return err
+			}
+
+			touchedIDs[profile.Name()] = struct{}{}
+
+			if err = ioutil.WriteFile(filepath.Join(ctrl.SeccompProfileDirectory, profile.Name()), value, 0o644); err != nil {
```

Comment:
as I understand it, seccomp profiles might be used by any time for containers, so writing a file like this might temporariliy render the data unusable (as file is truncated and re-written).

what we can do is to wrap this write into two-phase step: first check if the file already exists, if it does exist, try to read it; if the content matches, we can skip any writes to the file

Reply:
fixed, and added tests